compiler-core/src/javascript/typescript.rs

```diff
@@ -0,0 +1,530 @@
+            Statement::Fn { .. } => vec![],
+
+            Statement::ExternalFn {
+                public,
+                name,
+                arguments,
+                return_type,
+                ..
+            } if *public => vec![self.external_function(name, arguments, return_type)],
+            Statement::ExternalFn { .. } => vec![],
+        }
+    }
+
+    fn type_alias(&self, alias: String, type_: &Type) -> Output<'a> {
+        Ok(docvec![
+            "export type ",
+            Document::String(ts_safe_type_name(alias)),
+            " = ",
+            self.type_printer.print(type_),
+            ";"
+        ])
+    }
+
+    /// Converts a Gleam custom type definition into the Typescript equivalent.
+    /// In Gleam, all custom types have one to many concrete constructors. This
+    /// function first converts the constructors into Typescript then finally
+    /// emits a union type to represent the Typescript type itself. Because in
+    /// Gleam constructors can have the same name as the custom type, here we
+    /// append a "$" symbol to the emited Typescript type to prevent those
+    /// naming clases.
+    ///
+    fn custom_type_definition(
+        &mut self,
+        name: &str,
+        constructors: &'a [TypedRecordConstructor],
+    ) -> Vec<Output<'a>> {
+        let mut definitions: Vec<Output<'_>> = constructors
+            .iter()
+            .map(|constructor| Ok(self.record_definition(constructor)))
+            .collect();
+
+        definitions.push(Ok(docvec![
+            Document::String("export type ".to_string()),
+            Document::String(format!("{}$", name)),
+            Document::String(" = ".to_string()),
+            concat(Itertools::intersperse(
+                constructors.iter().map(|x| x.name.to_doc()),
+                break_("| ", " | "),
+            )),
+            Document::String(";".to_string()),
+        ]));
+
+        definitions
+    }
+
+    fn record_definition(&self, constructor: &'a TypedRecordConstructor) -> Document<'a> {
+        self.type_printer.set_prelude_used();
+        let head = docvec![
+            "export class ",
+            &constructor.name,
+            " extends $gleam.CustomType {"
+        ];
+
+        if constructor.arguments.is_empty() {
+            return head.append("}");
+        };
+
+        let parameters = concat(Itertools::intersperse(
+            constructor.arguments.iter().enumerate().map(|(i, arg)| {
+                let name = arg
+                    .label
+                    .as_ref()
+                    .map(|s| super::maybe_escape_identifier_doc(s))
+                    .unwrap_or_else(|| Document::String(format!("x{}", i)));
+                docvec![name, ": ", self.type_printer.print(&arg.type_)]
+            }),
+            break_(",", ", "),
+        ));
+
+        let class_body = docvec![line(), "constructor(", parameters, "): this",].nest(INDENT);
```

Comment:
I think constructors are not supposed to have any return annotation?

Reply:
Yeah, just looked again and you're totally right. Will also update this to wrap the argument list like the functions